microfs.py

```diff
@@ -56,10 +56,16 @@ def raw_on(serial):
     """
     Puts the device into raw mode.
     """
-    serial.write(b'\x03')  # Send CTRL-C to break out of loop.
-    serial.read_until(b'\n>')  # Flush buffer until prompt.
-    serial.write(b'\x01')  # Go into raw mode.
-    serial.read_until(b'\r\n>OK')  # Flush buffer until raw mode prompt.
+    attempts_left = 3
+    while attempts_left:
+        serial.write(b'\x03')    # Send CTRL-C to break out of loop.
+        if serial.read_until(b'\n>>>').endswith(b'\n>>>'):
```

Comment:
On Linux, this always returns `b''` so it never hits the `break` :-/

Reply:
Ctrl+B doesn't do anything on REPL mode, so perhaps the safes way is to do:
* Ctrl+B (ensures we are not in raw mode)
* Loop up to three times:
    * Ctrl+C 
    * read_until(b'\n>')

Reply:
The upstream pyboard.py script is very well tested and covers all cases.  Maybe try to copy the logic from there?

Reply:
@dpgeorge IIRC that's what I started from. ;-)

Reply:
For reference: https://github.com/micropython/micropython/blob/master/tools/pyboard.py#L286 right..?